### ISSUE TITLE
undeclared: gettimeofday

### ISSUE THREAD

**Issue by clausecker** [samtools/htscodecs#80]
When compiling on FreeBSD, `gettimeofday` is used without a declaration in some test files:

```
htscodecs/tests/rANS_static_test.c:148:5: warning: implicit declaration of function 'gettimeofday' is invalid in C99 [-Wimplicit-function-declaration]
    gettimeofday(&tv1, NULL);
    ^
1 warning generated.
htscodecs/tests/rANS_static4x16pr_test.c:148:5: warning: implicit declaration of function 'gettimeofday' is invalid in C99 [-Wimplicit-function-declaration]
    gettimeofday(&tv1, NULL);
    ^
1 warning generated.
htscodecs/tests/arith_dynamic_test.c:142:5: warning: implicit declaration of function 'gettimeofday' is invalid in C99 [-Wimplicit-function-declaration]
    gettimeofday(&tv1, NULL);
    ^
1 warning generated.
```

This can be fixed by including `<sys/time.h>` in these files.

**Comment by jmarshall** [comment]
If you look in those files you'll see they do in fact include `<sys/time.h>`.

On FreeBSD, this header provides a declaration of `gettimeofday()` [only if `__XSI_VISIBLE` is non-zero](https://cgit.freebsd.org/src/tree/sys/sys/time.h#n634). This is non-zero by default or if the code defines `_XOPEN_SOURCE` (see [cdefs.h](https://cgit.freebsd.org/src/tree/sys/sys/cdefs.h)), but will be reset to zero if you choose strict ANSI, C99, or C11 mode. Did you select any of those?

As recently predicted in https://github.com/samtools/htslib/issues/1586#issuecomment-1500144683 :smile:, HTSlib can ensure this is non-zero by ensuring `_XOPEN_SOURCE` is defined to an appropriate value. There's an old draft of that in [this xopen_source branch](https://github.com/jmarshall/htslib/compare/xopen_source) that probably needs some 500 vs 600 work (see also samtools/htslib#1246).

**Comment by clausecker** [comment]
I compiled htslib 1.17 with no changes on FreeBSD and tried to build and run the unit tests when this issue happened.

**Comment by clausecker** [comment]
The error occurs when building the unit tests (`make test`).  Did you try that?

I am building htslib 1.17 with clang.  We pass

    -O2 -pipe  -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing

as extra C flags when building this port.

**Comment by jmarshall** [comment]
It's a warning, not an error. Hence when these files are compiled via (HTSlib's build system's) `make test` the diagnostic messages scroll past quickly and are easily missed. (And the tests go on to run successfully anyway.)

FreeBSD's _cdefs.h_ also sidesteps the default activation of `__XSI_VISIBLE` [when `_POSIX_C_SOURCE` is set](https://cgit.freebsd.org/src/tree/sys/sys/cdefs.h#n704).

It turns out HTSlib's _Makefile_'s rules for _htscodecs/test/*.o_ add `‑D_POSIX_C_SOURCE=200112L` to their compilation. This is incorrect; it should be `‑D_XOPEN_SOURCE=600` instead, as these test source files use an XSI function as well as POSIX functionality.

That's the short-term fix. The better fix would be to set `_POSIX_SOURCE` globally in the HTSlib source code (as in the [xopen_source branch](https://github.com/jmarshall/htslib/compare/xopen_source) previously mentioned), but that would be something probably best done early in a development cycle.

**Comment by clausecker** [comment]
> It's a warning, not an error. Hence when these files are compiled via (HTSlib's build system's) make test the diagnostic messages scroll past quickly and are easily missed. (And the tests go on to run successfully anyway.)

Strictly speaking, calling an undeclared function causes behaviour to be undefined ever since C99.  It's just accepted as a legacy feature by recent compilers.  You should definitely fix this.

**Comment by clausecker** [comment]
@jkbonfield Yes, the is an htslibs issue.  I don't know why the issue was moved to this repository.

**Comment by jkbonfield** [comment]
Incorrect assumptions :)

I'll move it back.  Thanks for clarifying
